recirq/qcqmc/trial_wf.py

```diff
@@ -0,0 +1,1584 @@
+    return np.asarray(x)
+
+
+def _to_tuple(x: Iterable[LayerSpec]) -> Sequence[LayerSpec]:
+    return tuple(x)
+
+
+@attrs.frozen(repr=False)
+class PerfectPairingPlusTrialWavefunctionParams(TrialWavefunctionParams):
+    """Class for storing the parameters that specify a TrialWavefunctionData.
+
+    This class specifically stores the parameters for a trial wavefunction that
+    is a combination of a perfect pairing wavefunction with some number of
+    hardware-efficient layers appended.
+    """
+
+    name: str
+    hamiltonian_params: HamiltonianParams
+    heuristic_layers: Tuple[LayerSpec, ...] = attrs.field(converter=_to_tuple)
+    do_pp: bool = True
+    restricted: bool = False
+    random_parameter_scale: float = 1.0
+    n_optimization_restarts: int = 1
+    seed: int = 0
+    initial_orbital_rotation: Optional[np.ndarray] = attrs.field(
+        default=None, converter=lambda v: _to_numpy(v) if v is not None else None
+    )
+    initial_two_body_qchem_amplitudes: Optional[np.ndarray] = attrs.field(
+        default=None, converter=lambda v: _to_numpy(v) if v is not None else None
+    )
+    do_optimization: bool = True
+    use_fast_gradients: bool = False
+
+    @property
+    def n_orb(self) -> int:
+        return self.hamiltonian_params.n_orb
+
+    @property
+    def n_elec(self) -> int:
+        return self.hamiltonian_params.n_elec
+
+    @property
+    def n_qubits(self) -> int:
+        return 2 * self.n_orb
+
+    @property
+    def n_pairs(self) -> int:
+        return self.n_elec // 2
+
+    @property
+    def path_string(self) -> str:
+        return OUTDIRS.DEFAULT_TRIAL_WAVEFUNCTION_DIRECTORY + self.name
+
+    @property
+    def bitstrings(self) -> Iterable[Tuple[bool, ...]]:
+        return _get_bitstrings_a_b(n_orb=self.n_orb, n_elec=self.n_elec)
+
+    def _json_dict_(self):
+        # return cirq.dataclass_json_dict(self)
+        return attrs.asdict(self)
+
+    @property
+    def qubits_jordan_wigner_ordered(self) -> Tuple[cirq.GridQubit, ...]:
+        return _get_qubits_a_b(n_orb=self.n_orb)
+
+    @property
+    def qubits_linearly_connected(self) -> Tuple[cirq.GridQubit, ...]:
+        return _get_qubits_a_b_reversed(n_orb=self.n_orb)
+
+    @property
+    def mode_qubit_map(self) -> Dict[FermionicMode, cirq.GridQubit]:
+        return _get_mode_qubit_map_pp_plus(n_qubits=self.n_qubits)
+
+
+def _get_qubits_a_b(*, n_orb: int) -> Tuple[cirq.GridQubit, ...]:
+    # This ordering creates qubits to facilitate a Jordan Wigner string
+    # threading through a row of alpha orbitals in ascending order followed by a
+    # row of beta orbitals in ascending order.
+    return tuple(
+        [cirq.GridQubit(0, i) for i in range(n_orb)]
+        + [cirq.GridQubit(1, i) for i in range(n_orb)]
+    )
+
+
+def _get_qubits_a_b_reversed(*, n_orb: int) -> Tuple[cirq.GridQubit, ...]:
+    # This ordering creates qubits to facilitate operations that need a linearly
```

Comment:
I would change these line comments to triple quoted docstring.

Reply:
Done.